include/smack/smack.h

```diff
@@ -135,32 +287,146 @@ void __SMACK_decls() {
   D("axiom (forall i: int :: $fp2si($si2fp(i)) == i);");
   D("axiom (forall f: float :: $si2fp($fp2si(f)) == f);");
 
-  // Memory Model
-  D("function $base(int) returns (int);");
 
-  D("const unique $NULL: int;");
-  D("axiom $NULL == 0;");
+  // Memory Model
   D("const $UNDEF: int;");
+  D("function $base(ref) returns (ref);");
+  D("const unique $NULL: ref;");
+  D("const unique $UNIT: ref;");
+#ifdef BITVECTOR
+  //Pointer Arithmetic
+  D("function {:bvbuiltin \"bvadd\"} $add.ref(p1:ref, p2:ref) returns (ref);");
+  D("function {:bvbuiltin \"bvsub\"} $sub.ref(p1:ref, p2:ref) returns (ref);");
+  D("function {:bvbuiltin \"bvult\"} $ult.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvugt\"} $ugt.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvule\"} $ule.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvuge\"} $uge.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvslt\"} $slt.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvsgt\"} $sgt.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvsle\"} $sle.ref(p1:ref, p2:ref) returns (bool);");
+  D("function {:bvbuiltin \"bvsge\"} $sge.ref(p1:ref, p2:ref) returns (bool);");
+
+  D("function {:inline} $pa(pointer: ref, index: ref, size: ref) returns (ref) {$add.i32(pointer, $mul.i32(index, size))}");
+  D("function {:inline} $b2p(b: bool) returns (ref) {if b then $UNIT else $NULL}");
+
+#ifdef __x86_64__ 
```

Comment:
If for reason you wanted to minimize the number of constants, you could also refer to these as `$UNIT`, `$add.ref($UNIT,$UNIT)`, and so on.